cep-20.2.md

```diff
@@ -0,0 +1,831 @@
+# A new recipe format – part 2 - the allowed keys & values
+
+<table>
+<tr><td> Title </td><td> A new recipe format - values </td>
+<tr><td> Status </td><td> Proposed</td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
+<tr><td> Created </td><td> May 23, 2023</td></tr>
+<tr><td> Updated </td><td> May 23, 2023</td></tr>
+<tr><td> Discussion </td><td>  </td></tr>
+<tr><td> Implementation </td><td>https://github.com/prefix-dev/rattler-build</td></tr>
+</table>
+
+## Abstract
+
+We propose a new recipe format that is heavily inspired by conda-build. The main
+change is a pure YAML format without arbitrary Jinja or comments with semantic
+meaning.
+
+## Motivation
+
+The conda-build format is currently under-specified. For the new format, we try to list all values
+and types in a single document and document them. This is part of a larger effort (see [CEP-13](/cep-13.md) for the new YAML syntax).
+
+## History
+
+A discussion was started on what a new recipe spec could or should look like.
+The fragments of this discussion can be found here:
+https://github.com/mamba-org/conda-specs/blob/master/proposed_specs/recipe.md
+The reason for a new spec are:
+
+- Make it easier to parse ("pure yaml"). conda-build uses a mix of comments and
+  jinja to achieve a great deal of flexibility, but it's hard to parse the
+  recipe with a computer
+- iron out some inconsistencies around multiple outputs (build vs. build/script
+  and more)
+- remove any need for recursive parsing & solving
+- cater to needs for automation and dependency tree analysis via a determinstic
+  format
+
+## Major differences with conda-build
+
+Outputs and top-level package are mutually exclusive, and outputs have exactly the same structure as top-level keys.
+If outputs exist, the top-level keys are 'merged' with the output keys (e.g. for the about section).
+
+# Format
+
+## Schema version
+
+The implicit version of the YAML schema for a recipe is an integer 1.
+To discern between the "old" format and the new format, we utilize the file name.
+The old format is `meta.yaml` and the new format is `recipe.yaml`.
+The version can be explicitly set by adding a `schema_version` key to the recipe.
+
+```yaml
+# optional, since implicitly defaults to 1
+schema_version: 1 # integer
+```
+
+To benefit from autocompletion, and other LSP features in editors, we can add a schema URL to the recipe.
+
+```yaml
+# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
+```
+
+## Context section
+
+```yaml
+# The context dictionary defines arbitrary key-value pairs for Jinja interpolation
+# and replaces the {% set ... %} commands commonly used in recipes
+context:
+  variable: test
+  # note that we can reference previous values, that means that they are rendered in order
+  other_variable: test_${{ variable }}
```

Comment:
As already requested in the previous CEP:
If we want recipes to conform to the YAML specification (which I strongly suggest we should!), we cannot do this.

Please refer to https://yaml.org/spec/1.2.2/#3221-mapping-key-order :
> In the [representation](https://yaml.org/spec/1.2.2/#representation-graph) model, [mapping keys](https://yaml.org/spec/1.2.2/#nodes) do not have an order. To [serialize](https://yaml.org/spec/1.2.2/#serializing-the-representation-graph) a [mapping](https://yaml.org/spec/1.2.2/#mapping), it is necessary to impose an ordering on its [keys](https://yaml.org/spec/1.2.2/#nodes). This order is a [serialization detail](https://yaml.org/spec/1.2.2/#serializing-the-representation-graph) and should not be used when [composing](https://yaml.org/spec/1.2.2/#composing-the-representation-graph) the [representation graph](https://yaml.org/spec/1.2.2/#representation-graph) (and hence for the preservation of [application](https://yaml.org/spec/1.2.2/#processes-and-models) data). In every case where [node](https://yaml.org/spec/1.2.2/#nodes) order is significant, a [sequence](https://yaml.org/spec/1.2.2/#sequence) must be used. For example, an ordered [mapping](https://yaml.org/spec/1.2.2/#mapping) can be [represented](https://yaml.org/spec/1.2.2/#representation-graph) as a [sequence](https://yaml.org/spec/1.2.2/#sequence) of [mappings](https://yaml.org/spec/1.2.2/#mapping), where each [mapping](https://yaml.org/spec/1.2.2/#mapping) is a single [key/value pair](https://yaml.org/spec/1.2.2/#mapping). YAML provides convenient [compact notation](https://yaml.org/spec/1.2.2/#example-flow-mapping-adjacent-values) for this case.

To rephrase this for your proposal:
If you require some ordering in these mappings, then you're saying the recipes only conform to the specification of serialized YAML but not the YAML specification overall.
Which is not good since we're then again in a situation were we say "the recipes are YAML... but actually, only if you use certain tools to read/write them" (not as bad as "2 template formats on top one serialization format" as current `meta.yaml`, but related).

Meaning, this either has to be a list/sequence (e.g., `[{variable: test}, {other_variable: 'test_${{ variable }}'}]`) or require the consumer of these files to do topological sorting/cycle detection.

Reply:
I think this is no problem. We can easily create the "graph" from traversing the Jinja AST and finding the undefined variables. 

Most languages do have facilities to parse YAML into a Map that is sorted by insertion order (Rust and Python have it, for sure). So we can strongly encourage implementations to write (and read) this data in a sorted order.

Reply:
There is a difference between a format specification and implementation details.
If I do
```python
recipe = yaml.load("recipe.yaml")
recipe["context"] = dict(sorted(recipe["context"].items()))
yaml.dump(recipe, "recipe.yaml")`
```
then this does *not* change the YAML representation as per the YAML specification.

You **do not conform to theYAML specification** if you require this.

> So we can strongly encourage implementations

Yes, I'd also suggest that an implementation should not change the order (having things stable causes less confusion and possible determinism).
But that does not mean the implementations should *have to* do that.
Otherwise, please don't name these files `.yaml`.

To reiterate: This is a specification bug that was already listed amongst others in https://github.com/conda-incubator/ceps/pull/54#pullrequestreview-1515301216 .

Reply:
Sure. We _can_ topologically sort the context object. So we can get rid of this requirement (but still ask implementations to do it because it will be much more logical for humans).

Reply:
I added a note with more details and the request to implementations to do topological sorting.